### ISSUE TITLE
zkasm: implement memory access operations and memory sections

### ISSUE THREAD

**Issue by nagisa** [near/wasmtime#110]
We need to have the following wasm instructions implemented, initially and primarily the specific instructions used by [this benchmark](https://github.com/near/wasmtime/issues/90#issuecomment-1825561270)

* [ ] `iN.store memarg`
* [ ] `iN.load memarg`
* [ ] `iN.storeM memarg`
* [ ] `iN.loadM_sx memarg`

all of these instructions allow for a compile-constant `memarg` immediate to be specified, where `memarg` can contain an address offset and the expected alignment (expressed as the exponent of a power of 2).

So 

```
i32.const 1000
iN.load offset=1234 align=4
```

really says that the zkasm module ought to load a value at the address of `1234 + 1000`, and that the address is *anticipated* to be aligned to `2^4 = 16` bytes. Note that `2234 % 16` is not 0 and so the load is not actually aligned, but this instruction is still perfectly valid according to the WASM specification. I anticipate that we should ignore the `align` annotation altogether and that it is largely useful for JIT purposes which can make runtime decisions based on whether the addresses do generally turn out to be aligned according to this specification or not.

So all of the loads and stores we make will have to assume alignment of 1, which `zkasm` does not innately support. In order to emulate unaligned loads we will have to implement load/store tearing ourselves. For loads and stores that modify a memory slot-sized memory region (64-bits?):

```
  REQUIRED_ALIGN_MASK = 0b11..1 // required bytes of alignment minus one
  COMPUTED_ADDRESS = a + offset
  ALIGNED_ADDRESS = COMPUTED_ADDRESS & ~REQUIRED_ALIGN_MASK;
  MISALIGNED_OFFSET = MISALIGNED_OFFSET & REQUIRED_ALIGN_MASK;
  MEMORY_SLOT = convert_to_slot(ALIGNED_ADDRESS)
  A: MLOAD(MEMORY_SLOT)
  MISALIGNED_OFFSET: JMPZ(load_complete)
  B: MLOAD(MEMORY_SLOT + 1)
  ${A << MISALIGNED_OFFSET | B >> (ALIGNMENT_BYTES - MISALIGNED_OFFSET)} => A
load_complete:
...
```

or alternatively check if `MEM_ALIGN_RD` (and `MEM_ALIGN_WR`) instructions implement the same behaviour for us (the documentation claims that they do, but after our discussion with polygon folks I ended up understanding that these instructions do not work for the purpose…?) Anyway, I think something like this should work:

```
    $ => A: MLOAD(MEMORY_SLOT)
    MISALIGNED_OFFSET: JMPZ(load_complete)
    $ => B: MLOAD(MEMORY_SLOT + 1)
    MISALIGNED_OFFSET => C
    $ => A: MEM_ALIGN_RD
load_complete:
```

With this all written down, I think the codegen part of of 64-bit memory accesses is quite straightforward.

## Sub-slot sized loads (and stores!)

In order to load a 32-bit integer/16-bit/8-bit values with instructions that inherently work with 64-bits, additional masking is required. Fairly straightforward with an `A & iN::max_value()` after `load_complete`. Stores, however need to retain surrounding bits, so those will need to load the value, modify it and write the modified result back.

`MEM_ALIGN_WR8` has a specialized implementation for writing out a byte, but for 16-bit and 32-bit writes we need to do our own thing (perhaps polygon people would agree to adding `MEM_ALIGN_WR16` and `MEM_ALIGN_WR32`?) For 32-bits:

```
i64.load // replicate the `MEM_ALIGN_RD` based memory load sequence here
i64.and (i64.const 0xFFFF_FFFF_0000_0000) // clear out the low 32-bits
i64.or // set the low 32-bits with the new value
i64.store // this is now unaligned full-slot store, which is implementable with `MEM_ALIGN_WR`.
```

(This is an example where using ISLE-based lowering rules might make it simpler to encode this implementation)

## Memory intialization?

WASM has this [data segment](https://www.w3.org/TR/wasm-core-1/#data-segments%E2%91%A0) that looks like `(data "12345")` in your average `.wat` files. This data is expected to be populated at the supplied offsets as the program is loaded into memory. However zkASM does not have a way to encode this concept into the binary, so this initial data is going to have to make its way to the prover by some other means. For example we could write out a separate `.bin` file that the prover loads and initializes the zkASM memory with before it starts executing the code.

**Comment by nagisa** [comment]
On the cranelift side these load instructions show up as `sload8`, `uload{N}`, `iload{N}` and `load` for 64-bit loads. Note in particular that all of these load into a 64-bit register (nice, just what we need) and `s`, `u` and `i` prefixes basically just describe whether the loaded value needs to be sign-extended or not.

## Sign extension

Another interesting problem to solve. Imagine you’re loading an 8-bit value that is to be interpreted as a signed number. So we're loading something like `-2`, which is `0xfe` in binary. When loaded as-is into a 64-bit register, without sign extension, you will end up a `254` instead, which is not quite what the intent here is. The value should still be `-2`, which in binary is `0xFFFF_FFFF_FFFF_FFFe`!

The sign extension is actually quite straightforward thing to reason about – the most significant sign bit (of the 8-bit value in this example) is duplicated to the higher bits of the register. So if you need to sign extend a 8-bit value `0b1000_0000` to a 64-bit one, you get `0b11...1000_0000`. If the sign bit is zero as in `0b0100_0000`, then sign-extended you will get `0b00...0100_000`.

zkASM does not appear to have an instruction for this as far as I can tell… I'll have to think about how to best replicate this…

**Comment by nagisa** [comment]
> From reading the [ZK ASM architecture doc](https://github.com/0xPolygonHermez/zkevm-techdocs/blob/main/zkevm-architecture/v.1.1/zkevm-architecture.pdf) on page 25 I see that there is already an MSTORE8 EVM opcode [implementation](https://github.com/0xPolygonHermez/zkevm-rom/blob/877328eff50070426f040a8843b7d6eb3140ef22/main/opcodes/storage-memory.zkasm#L81) that is actually [cheaper](https://github.com/0xPolygonHermez/zkevm-rom/blob/877328eff50070426f040a8843b7d6eb3140ef22/docs/opcode-cost-zk-counters.md?plain=1#L59) than MSTORE in the terms of counters.

Yeah, it is using the `MEM_ALIGN_WR8` opcode I mentioned above, but there isn’t one such for 16-bit and 32-bit operations. Worth asking if the intent is to open-code these wider operations or do something else.

**Comment by akashin** [comment]
> Yeah, it is using the `MEM_ALIGN_WR8` opcode I mentioned above, but there isn’t one such for 16-bit and 32-bit operations. Worth asking if the intent is to open-code these wider operations or do something else.

We discussed this at a recent sync, summary:
- It would be easy to extend MEM_ALIGN_WR to other bitwidths if we need to
- In PIL2 MEM_ALIGN will likely be parameterizable by bitwidth
- If there are multiple MEM_ALIGN_WR8 in a row, there are ways to represent this with less instructions, we can get help from Polygon on how exactly this can be done when we have an example ZK ASM file to optimize

I think for now we can proceed by using less optimized MEM_ALIGN_WR8 and MEM_ALIGN_RD and when PIL2 comes along, we will work on a more optimized version.

**Comment by nagisa** [comment]
Did you have a chance to discuss sign extension? Though now that I think about it, it isn’t a problem we need to solve for the sha256 benchmark as it deals with unsigned everywhere.

**Comment by akashin** [comment]
> Did you have a chance to discuss sign extension? Though now that I think about it, it isn’t a problem we need to solve for the sha256 benchmark as it deals with unsigned everywhere.

No, we didn't discuss the sign extension - I've added it as a topic for discussion next time.
I agree that we likely don't need it in the near future.
